Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift

```diff
@@ -9,201 +9,1069 @@
+                preconditionFailure("Invalid state")
+
+            case let .merging(task, buffer, upstreamContinuations, upstreamsFinished, .some(downstreamContinuation)):
+                // We produced an element and have an outstanding downstream continuation
+                // this means we can go right ahead and resume the continuation with that element
+                precondition(buffer.isEmpty, "We are holding a continuation so the buffer must be empty")
+
+                state = .merging(
+                    task: task,
+                    buffer: buffer,
+                    upstreamContinuations: upstreamContinuations,
+                    upstreamsFinished: upstreamsFinished,
+                    downstreamContinuation: nil
+                )
+
+                return .resumeContinuation(
+                    downstreamContinuation: downstreamContinuation,
+                    element: element
+                )
+
+            case .merging(let task, var buffer, let upstreamContinuations, let upstreamsFinished, .none):
+                // There is not outstanding downstream continuation so we must buffer the element
+                // This happens if we race our upstream sequences to produce elements
+                // and the _losers_ are signalling their produced element
+                state = .modifying
+
+                buffer.append(element)
+
+                state = .merging(
+                    task: task,
+                    buffer: buffer,
+                    upstreamContinuations: upstreamContinuations,
+                    upstreamsFinished: upstreamsFinished,
+                    downstreamContinuation: nil
+                )
+
+                return .none
+
+            case .upstreamThrew:
+                // Another upstream already produced an error so we just drop the new element
+                return .none
+
+            case .finished:
+                // Since cancellation is cooperative it might be that child tasks
+                // are still producing elements after we finished.
+                // We are just going to drop them since there is nothing we can do
+                return .none
+
+            case .modifying:
+                preconditionFailure("Invalid state")
+            }
+        }
+
+        /// Actions returned by `upstreamFinished()`.
+        enum UpstreamFinishedAction {
+            /// Indicates that the task and the upstream continuations should be cancelled.
+            case cancelTaskAndUpstreamContinuations(
+                task: Task<Void, Never>,
+                upstreamContinuations: [CheckedContinuation<Void, Error>]
+            )
+            /// Indicates that the downstream continuation should be resumed with `nil` and
+            /// the task and the upstream continuations should be cancelled.
+            case resumeContinuationWithNilAndCancelTaskAndUpstreamContinuations(
```

Comment:
some of these long names and numerous associated types kinda almost infer structness of the associations. Would it be better to change them to structures instead of effectively tuples?

Reply:
In general I am with you; however, this would introduce just a struct for one case of an action. My personal opinion, is that this is overkill for something internal. If you this is important to you, I am happy to change it though. Your call :)